pkg/asset/installconfig/gcp/usertags.go

```diff
@@ -0,0 +1,298 @@
+func (p *processedUserTags) copy() map[string]string {
+	p.Lock()
+	defer p.Unlock()
+
+	t := make(map[string]string, len(p.tags))
+	for k, v := range p.tags {
+		t[k] = v
+	}
+	return t
+}
+
+// NewTagManager creates a tagManager instance.
+func NewTagManager(client API) TagManager {
+	return &tagManager{client: client}
+}
+
+// GetUserTags returns the processed list of user provided tags if already available,
+// else validates, stores and returns the processed tags.
+func GetUserTags(mgr TagManager, ic *types.InstallConfig) (map[string]string, error) {
+	return getProcessedTags(mgr, ic)
+}
+
+// getProcessedTags returns the list user-provided validated tags.
+func getProcessedTags(mgr TagManager, ic *types.InstallConfig) (map[string]string, error) {
+	if !processedTags.getState() {
+		if err := validateUserTags(mgr, ic); err != nil {
+			return nil, err
+		}
+	}
+
+	return processedTags.copy(), nil
+}
+
+// validateAndStoreUserTags validates user provided tags are accessible and exists.
+// And converts and stores tags from user provided list of tags in NamespacedName
+// `{parentID}/{tag_key_short_name}/{tag_value_short_name}` form to key:value pairs,
+// with key of the form `tagKeys/{tag_key_id}` and value of the form
+// `tagValues/{tag_value_id}`. Returns error when fetching a tag detail fails or when
+// tag already exists on the project resource.
+func validateAndStoreUserTags(ctx context.Context, mgr TagManager, project string, userTags []gcp.UserTag) error {
+	if len(userTags) == 0 {
+		return nil
+	}
+
+	start := time.Now()
+	logrus.Debugf("user defined tags list: %v", userTags)
+
+	if len(userTags) > maxUserTagLimit {
+		return fmt.Errorf("more than %d user tags is not allowed, configured count: %d", maxUserTagLimit, len(userTags))
+	}
+
+	projectTags, err := getProjectTags(context.TODO(), mgr, project)
+	if err != nil {
+		return err
+	}
+
+	if dupTags := findDuplicateTags(userTags, projectTags); len(dupTags) != 0 {
+		return fmt.Errorf("found duplicate tags, %v tags already exist on %s project", dupTags, project)
+	}
+
+	client, err := getTagValuesClient(ctx, mgr)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	processedTags.make(len(userTags))
+	nonexistentTags := make([]string, 0)
+	for _, tag := range userTags {
+		name := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		key, value, err := client.getTagKeyValueName(ctx, name)
+		if err != nil {
+			// check and return all non-existing tags at once
+			// for user to fix in one go.
+			var gErr *apierror.APIError
+			// google API returns StatusForbidden or StatusNotFound when the tag
+			// does not exist, since it could be because of permission issues
+			// or genuinely tag does not exist.
+			if errors.As(err, &gErr) && (gErr.HTTPCode() == http.StatusNotFound ||
+				gErr.HTTPCode() == http.StatusForbidden) {
+				logrus.Debugf("does not have permission to access %s tag or does not exist", name)
```

Comment:
Error returned by the GCP APIs is also similar. I think it's because RBAC can be defined on each tag key and value, and with a given service account, it won't be possible to say whether the tag exists and is inaccessible or it doesn't exist at all, so thought to keep the same.

```
User [xyz] does not have permission to access tagValues instance [namespaced] (or it may not exist): Permission denied on resource 'openshift/key/value' (or it may not exist).
```

Reply:
Code updated, new API of `cloudresourcemanager` package returns `StatusForbidden` for both non-existent and inaccessible tags.